rust/src/operations/optimize.rs

```diff
@@ -641,61 +651,86 @@ impl MergePlan {
 
                         let object_store = object_store.clone();
 
                         let rewrite_result = tokio::task::spawn(Self::rewrite_files(
-                            self.task_parameters.clone(),
+                            task_parameters.clone(),
                             partition,
                             files,
                             object_store,
                             batch_stream,
                         ));
                         util::flatten_join_error(rewrite_result)
                     })
-                    .buffer_unordered(self.max_concurrent_tasks)
-                    .try_for_each(|(partial_actions, partial_metrics)| {
-                        debug!("Recording metrics for a completed partition");
-                        actions.extend(partial_actions);
-                        metrics.add(&partial_metrics);
-                        async { Ok(()) }
-                    })
-                    .await?;
+                    .boxed()
             }
-        }
+        };
 
-        metrics.preserve_insertion_order = true;
-        if metrics.num_files_added == 0 {
-            metrics.files_added.min = 0;
-        }
-        if metrics.num_files_removed == 0 {
-            metrics.files_removed.min = 0;
-        }
+        let mut stream = stream.buffer_unordered(self.max_concurrent_tasks);
 
-        // TODO: Check for remove actions on optimized partitions. If a
-        // optimized partition was updated then abort the commit. Requires (#593).
-        if !actions.is_empty() {
-            let mut metadata = Map::new();
-            metadata.insert("readVersion".to_owned(), self.read_table_version.into());
-            let maybe_map_metrics = serde_json::to_value(metrics.clone());
-            if let Ok(map) = maybe_map_metrics {
-                metadata.insert("operationMetrics".to_owned(), map);
+        let mut table = DeltaTable::new_with_state(object_store.clone(), snapshot.clone());
+
+        let mut last_commit = Instant::now();
+        loop {
+            let next = stream.next().await.transpose()?;
+
+            let end = next.is_none();
+
+            if let Some((partial_actions, partial_metrics)) = next {
+                debug!("Recording metrics for a completed partition");
+                actions.extend(partial_actions);
+                metrics.add(&partial_metrics);
+                total_metrics.add(&partial_metrics);
             }
 
-            commit(
-                object_store.as_ref(),
-                &actions,
-                self.task_parameters.input_parameters.clone().into(),
-                snapshot,
-                Some(metadata),
-            )
-            .await?;
+            let now = Instant::now();
+            if !actions.is_empty() && (self.min_commit_interval.map_or(false, |i| now.duration_since(last_commit) > i) || end) {
+                let actions = std::mem::take(&mut actions);
+                last_commit = now;
```

Comment:
Does it matter if this is updated before the commit success?

Reply:
I don't think it matters much, because `min_commit_interval` is supposed to be much larger than the time it takes to do a commit.